File: leetcode/javascript/stacksAndQueues/movingAverageFromDataStream(easy)/movingAverageFromDataStream.js
/* 346. Moving Average from Data Stream

Учитывая поток целых чисел и размер окна, вычислите скользящее среднее всех целых чисел в скользящем окне.

* For example,
*
* MovingAverage m = new MovingAverage(3);
* m.next(1) = 1
* m.next(10) = (1 + 10) / 2
* m.next(3) = (1 + 10 + 3) / 3
* m.next(5) = (10 + 3 + 5) / 3

Input: ["MovingAverage", "next", "next", "next", "next"]
[[3], [1], [10], [3], [5]]
Output: [null, 1.0, 5.5, 4.66667, 6.0]
*/

/**
 * @param {number} size
 */
var MovingAverage = function(size) {
    this.n = size;
    this.queue = [];
    this.average = 0.0;
};
/**
 * @param {number} val
 * @return {number}
 */
MovingAverage.prototype.next = function(val) {
    let removedVal;

    if (this.queue.length >= this.n) {
        removedVal = this.queue.shift();
        this.average = this.average - removedVal;
    }
    this.queue.push(val);
    this.average += val;

    return this.average / this.queue.length;
};

const movingAverage = new MovingAverage(3);

movingAverage.next(1); // return 1.0 = 1 / 1
movingAverage.next(10); // return 5.5 = (1 + 10) / 2
movingAverage.next(3); // return 4.66667 = (1 + 10 + 3) / 3
movingAverage.next(5); // return 6.0 = (10 + 3 + 5) / 3

File: leetcode/go/stacksAndQueues/movingAverageFromDataStream(easy)/movingAverageFromDataStream.go
package main

/* 346. Moving Average from Data Stream

Учитывая поток целых чисел и размер окна, вычислите скользящее среднее всех целых чисел в скользящем окне.

* For example,
*
* MovingAverage m = new MovingAverage(3);
* m.next(1) = 1
* m.next(10) = (1 + 10) / 2
* m.next(3) = (1 + 10 + 3) / 3
* m.next(5) = (10 + 3 + 5) / 3

Input: ["MovingAverage", "next", "next", "next", "next"]
[[3], [1], [10], [3], [5]]
Output: [null, 1.0, 5.5, 4.66667, 6.0]
*/

func main() {
	movingAverage := Constructor(3)

	println(movingAverage.Next(1))  // 1.0
	println(movingAverage.Next(10)) // 5.5
	println(movingAverage.Next(3))  // 4.666666666666667
	println(movingAverage.Next(5))  // 6.0
}

type MovingAverage struct {
	size  int
	queue []int
	sum   float64
}

func Constructor(size int) MovingAverage {
	return MovingAverage{
		size:  size,
		queue: make([]int, 0, size),
		sum:   0.0,
	}
}

func (this *MovingAverage) Next(val int) float64 {
	if len(this.queue) >= this.size {
		removedVal := this.queue[0]
		this.queue = this.queue[1:]
		this.sum -= float64(removedVal)
	}

	this.queue = append(this.queue, val)
	this.sum += float64(val)

	return this.sum / float64(len(this.queue))
}
